### ISSUE TITLE
Weird interaction between text-rotation-alignment and text-translate-anchor

### ISSUE THREAD

**Issue by kubapelc** [maplibre/maplibre-gl-js#3456]
Hi,

while studying the behaviour of `text-translate`, `text-translate-anchor` and `text-rotation-alignment` I noticed some weird interactions between those style tags. Below is a summary of their interactions. Term "map space" refers to text pitch aligned with map (flat text as if printed onto the map), "screen space" is text with pitch aligned with viewport, so always flat on the screen. The described behaviour applies to both `symbol-placement` set to `point` and `line`.

| text-rotation-alignment | translate-anchor | current behaviour                                                     | note       | suggested behaviour                             |
| ----------------------- | ---------------- | --------------------------------------------------------------------- | ---------- | ----------------------------------------------- |
| viewport                | map              | viewport-rotated translation done in map space, screen space text     | reasonable | unmodified |
| viewport                | viewport         | viewport-aligned screen space translation, screen space text           | reasonable | unmodified |
| map                     | map              | map space translation rotated in *opposite* direction relative to viewport rotation, text in map space | broken?    | correct map-aligned rotation, text in map space |
| map                     | viewport         | map-aligned rotation of text translation, text translated in map space, map space text                  | weird?     | text translated in map space in direction that is *viewport*-rotated, text rendered in map space |

See these demos below, try to rotate & pitch the camera. The first one is rotating in the wrong direction, the second one should rotate, but doesn't (and also it exhibits some weird stretching).

~~https://kvaleya.gitlab.io/maplibre/aligndemo/rotation-align-map-translate-anchor-map.html~~

~~https://kvaleya.gitlab.io/maplibre/aligndemo/rotation-align-map-translate-anchor-viewport.html~~

Better demo: https://kvaleya.gitlab.io/maplibre/aligndemo/combineddemo.html

I'm currently modifying the symbol rendering code for globe and I am kinda familiar with it, so I can probably make a PR with a fix fairly quickly, but I'm not sure what the intended behaviour is.

Is this unintended behaviour? Should I make a PR with these suggested changes?

**Comment by kubapelc** [comment]
I've played around with this for a bit and made a better demo:

https://kvaleya.gitlab.io/maplibre/aligndemo/combineddemo.html

Now all 4 variants seem inconsistent with how `-translate` and `-translate-anchor` works for any other layer.

**Comment by HarelM** [comment]
I think the red and black dot in the demo are doing a fine job at explaining the problem.
I tend to agree about your conclusions.
Seems to me like map-map (third point from the left) is doing twice the rotation instead of once.
I would expect the last point (the one on the right) to rotate the text around the red dot basically.

Would be interesting to see how this behaves in native.
@louwers is there a quick way someone could test how this is behaving in native? an easy way to run some render test maybe?

Should this issue be part of the style spec repo maybe (should I transfer it)?

**Comment by louwers** [comment]
I missed your previous comment. Is it still relevant?

I am happy to run a render test if I can get a `style.json` for it.

**Comment by HarelM** [comment]
The style is in the demo page above, it has a geojson with some points. You should be able to extract the style/data from it:
```js

		const originE = -116.231;
		const originN = 43.604 - 0.35;
		const map = new maplibregl.Map({
			container: 'map',
			style:
				`https://api.maptiler.com/maps/streets/style.json?key=${maptiler_apikey}`,
			center: [originE + 0.003, originN],
			zoom: 17
		});

		map.on('load', () => {
			//
			// Labels
			//
			map.addSource("maintext", {
				'type': 'geojson',
				'data': {
					'type': 'Feature',
					'geometry': {
						'type': 'Point',
						'coordinates': [originE + 0.003, originN + 0.001]
					}
				}
			});
			map.addLayer({
				'id': "maintext",
				'type': 'symbol',
				'source': "maintext",
				'layout': {
					'text-field': [
						'format',
						"Black dots are the original anchor points - this is where the text resides.\nRed dots are anchors translated in the same manner as the text's layers.\nNote how no text seems to match its dot.\nCollision boxes are also mismatched.\nRotate and pitch the map to see this in action.", {}
					],
					'text-font': ['Open Sans Semibold', 'Arial Unicode MS Bold'],
					'text-offset': [0, 0.6],
					'text-anchor': 'top',
					'text-rotation-alignment': "map"
				},
			});

			//
			// Alignment tests
			//
			for (let i = 0; i < 4; i++) {
				map.addSource(`test${i}`, {
					'type': 'geojson',
					'data': {
						'type': 'Feature',
						'geometry': {
							'type': 'Point',
							'coordinates': [originE + i * 0.002, originN]
						}
					}
				});
				const translate = i % 2 == 0 ? "map" : "viewport";
				const rotate = ((i / 2) >> 0) !== 0 ? "map" : "viewport";
				map.addLayer({
					'id': `testLayer${i}`,
					'type': 'symbol',
					'source': `test${i}`,
					'layout': {
						'text-field': [
							'format',
							`text-rotation-alignment: ${rotate}\ntranslate-anchor: ${translate}`, {}
						],
						'text-font': ['Open Sans Semibold', 'Arial Unicode MS Bold'],
						'text-offset': [0, 0.6],
						'text-anchor': 'top',
						'text-rotation-alignment': rotate
					},
					'paint': {
						'text-translate': [0, 150],
						'text-translate-anchor': translate
					}
				});
				map.addLayer({
					'id': `testLayer${i}-circle`,
					'type': 'circle',
					'source': `test${i}`,
				});
				map.addLayer({
					'id': `testLayer${i}-circle2`,
					'type': 'circle',
					'source': `test${i}`,
					'paint': {
						'circle-translate': [0, 150],
						'circle-translate-anchor': translate,
						'circle-color': "#ff0000",
					}
				});
			}
			map.showCollisionBoxes = true;
		});
	
```

**Comment by kubapelc** [comment]
> Is it possible to only open a PR with this fix only against `main` so that we can merge that regadless of the globe progress?

@HarelM Sadly I think it would be impractical to do a separate PR now since I've made a lot of changes around symbol rendering files, and this fix relies on them. I've also found out that it still has a lot of issues, such as variable anchors not working (see https://github.com/maplibre/maplibre-gl-js/issues/210 ) and lots of other more or less subtle problems. MapTiler maps might look ok, but a lot of symbol related render tests fail in my branch.
